syntax = "proto3";

package amanda;

message Value {
  oneof kind {
    Struct struct_value = 1;
    ListValue list_value = 2;
    SerializedValue serialized_value = 3;
    NullValue null_value = 4;
    bool bool_value = 5;
    int64 int_value = 6;
    double double_value = 7;
    string string_value = 8;
    bytes bytes_value = 9;
    DataTypeValue type_value = 10;
    RefValue ref_value = 11;
  }
}

message DataTypeValue {
  string namespace = 1;
  string name = 2;
  Struct attrs = 3;
}

message Struct {
  map<string, Value> fields = 1;
}

enum NullValue {
  NULL_VALUE = 0;
}

message ListValue {
  repeated Value values = 1;
}

message SerializedValue {
  DataTypeValue type = 1;
  Value value = 2;
}

message RefValue {
  string schema = 1;
  string value = 2;
}
